### PR TITLE
Adds recentering task script and methods, with ctests

### LINKED ISSUE #912: Recenter ens. members around the deterministic background.
The code/application exists (under utils) but is not implemented in the workflow yet. We need to pass recentering increments to the ensemble members.

### PR DESCRIPTION
**Pull request by AndrewEichmann-NOAA**
Resolves https://github.com/NOAA-EMC/GDASApp/issues/912

### CODE REVIEW

File: parm/soca/berror/soca_ensrecenter.yaml [hunk 1]
```diff
@@ -0,0 +1,102 @@
+# This yaml is for applying deterministic recentering increments to the ensemble members
+
+geometry:
+  mom6_input_nml: mom_input.nml
+  fields metadata: ./fields_metadata.yaml
+
+date: '{{ATM_WINDOW_BEGIN}}'
```
Comment (CoryMartin-NOAA) [inline parm/soca/berror/soca_ensrecenter.yaml]:
```suggestion
date: '{{ ATM_WINDOW_BEGIN }}'
```

File: parm/soca/berror/soca_ensrecenter.yaml [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+layers variable: [hocn]
+
+increment variables: [tocn, socn, uocn, vocn, ssh, hocn, cicen, hicen, hsnon]
+
+set increment variables to zero: [ssh]
+
+vertical geometry:
+  date: '{{ATM_WINDOW_BEGIN}}'
```
Comment (CoryMartin-NOAA) [inline parm/soca/berror/soca_ensrecenter.yaml]:
```suggestion
  date: '{{ ATM_WINDOW_BEGIN }}'
```

File: parm/soca/berror/soca_ensrecenter.yaml [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  basename: ./INPUT/
+  ocn_filename: MOM.res.nc
+  read_from_file: 3
+
+add recentering increment: false
+
+soca increments:  # Could also be states, but they are read as increments
+  number of increments: 4 $${ENS_SIZE}
```
Comment (CoryMartin-NOAA) [inline parm/soca/berror/soca_ensrecenter.yaml]: what's with this key? shouldn't it be `{{ ENS_SIZE }}` which should really be `{{ NMEM }}`?

File: parm/soca/berror/soca_ensrecenter.yaml [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  pattern: '%mem%'
+  template:
+    date: '{{ATM_WINDOW_BEGIN}}'
```
Comment (CoryMartin-NOAA) [inline parm/soca/berror/soca_ensrecenter.yaml]:
```suggestion
    date: '{{ ATM_WINDOW_BEGIN }}'
```

File: parm/soca/berror/soca_ensrecenter.yaml [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+    ice_filename: 'ice.%mem%.nc'
+    read_from_file: 3
+
+steric height:
+  linear variable changes:
+  - linear variable change name: BalanceSOCA  # Only the steric balance is applied
+
+#ensemble mean output:
+#  datadir: ./static_ens
+#  date: '{{ATM_WINDOW_BEGIN}}'
```
Comment (CoryMartin-NOAA) [inline parm/soca/berror/soca_ensrecenter.yaml]: I'm going to stop from here on out, but all of these should have spaces between the bracket and variable name per Jinja2 standard

File: parm/templates/recen.yaml [hunk 6]
```diff
@@ -0,0 +1,12 @@
+window begin: '{{window_begin}}'
+window length: $(window_length)
+background_dir: '{{background_dir}}'
+staticb_dir: !ENV ${STATICB_DIR}
```
Comment (CoryMartin-NOAA) [inline parm/templates/recen.yaml]: We should get rid of all of these !ENV, $(), and use the standard Jinja2 everywhere now

File: ush/soca/marine_recenter.py [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+
+        half_assim_freq = timedelta(hours=int(config['assim_freq'])/2)
+        window_begin = cdate - half_assim_freq
+        window_begin_iso = window_begin.strftime('%Y-%m-%dT%H:%M:%SZ')
+        window_middle_iso = cdate.strftime('%Y-%m-%dT%H:%M:%SZ')
+
+        self.window_config = {'window_begin': f"{window_begin.strftime('%Y-%m-%dT%H:%M:%SZ')}",
+                              'ATM_WINDOW_BEGIN': window_begin_iso,
+                              'ATM_WINDOW_MIDDLE': window_middle_iso,
+                              'ENS_SIZE': self.config.NMEM_ENS,
```
Comment (CoryMartin-NOAA) [inline ush/soca/marine_recenter.py]: yeah like I said above, we shouldn't rename variables, and it should be NMEM_ENS everywhere.

File: ush/soca/marine_recenter.py [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+            for domain in ['ocean', 'ice']:
+                mem_dir = os.path.join(self.config.ROTDIR,
+                                      f'enkf{self.runtime_config.RUN}.{gPDYstr}',
+                                      f'{self.runtime_config.gcyc}',
+                                      f'mem{str(mem).zfill(3)}',
+                                      'model_data',
+                                      domain,
+                                      'history')
+                mem_dir_real = os.path.realpath(mem_dir)
+                f009 = f'enkfgdas.{domain}.t{self.runtime_config.gcyc}z.inst.f009.nc'
```
Comment (CoryMartin-NOAA) [inline ush/soca/marine_recenter.py]: bad! what about `enkfgfs`?

File: ush/soca/marine_recenter.py [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+
+                fname_in = os.path.abspath(os.path.join(mem_dir_real, f009))
+                fname_out = os.path.realpath(os.path.join(self.config.ens_dir,
+                                             domain+"."+str(mem)+".nc"))
+                ens_member_list.append([fname_in, fname_out])
+        # TODO(AFE) any reason not to make this a link?
```
Comment (CoryMartin-NOAA) [inline ush/soca/marine_recenter.py]: NCO wants all copied, no links

**Review by guillaumevernieres — COMMENTED** [review]
Looks good to me @AndrewEichmann-NOAA . We'll trigger the `ci` after you address @CoryMartin-NOAA 's comments and fix the norms.
I'm pretty sure the re-centering isn't correct yet ... But that's my fault. We'll sort out the configuration details in a subsequent PR.

File: ush/soca/marine_recenter.py [hunk 10]
```diff
@@ -0,0 +1,248 @@
...
+        FileHandler({'copy': ens_member_list}).sync()
+
+        ################################################################################
+        # generate the YAML file for the post processing of the clim. ens. B
```
Comment (guillaumevernieres) [inline ush/soca/marine_recenter.py]: that's not what's done here

File: ush/soca/marine_recenter.py [hunk 11]
```diff
@@ -0,0 +1,248 @@
...
+
+        logger.info(f"---------------- generate soca_ensb.yaml")
+
+        berr_yaml = YAMLFile(path=self.config.berr_yaml_template)
```
Comment (guillaumevernieres) [inline ush/soca/marine_recenter.py]: bad choice a variable name, use something adequately descriptive

File: parm/soca/berror/soca_ensrecenter.yaml (hunk 3 above)
Comment (AndrewEichmann-NOAA) [reply parm/soca/berror/soca_ensrecenter.yaml]: changed

File: parm/soca/berror/soca_ensrecenter.yaml (hunk 4 above)
Comment (AndrewEichmann-NOAA) [reply parm/soca/berror/soca_ensrecenter.yaml]: changed

File: parm/soca/berror/soca_ensrecenter.yaml (hunk 5 above)
Comment (AndrewEichmann-NOAA) [reply parm/soca/berror/soca_ensrecenter.yaml]: done

File: ush/soca/marine_recenter.py (hunk 11 above)
Comment (AndrewEichmann-NOAA) [reply ush/soca/marine_recenter.py]: changed

File: ush/soca/marine_recenter.py (hunk 10 above)
Comment (AndrewEichmann-NOAA) [reply ush/soca/marine_recenter.py]: corrected

File: ush/soca/marine_recenter.py (hunk 9 above)
Comment (AndrewEichmann-NOAA) [reply ush/soca/marine_recenter.py]: comment removed

File: ush/soca/marine_recenter.py (hunk 7 above)
Comment (AndrewEichmann-NOAA) [reply ush/soca/marine_recenter.py]: changed

File: parm/soca/berror/soca_ensrecenter.yaml (hunk 2 above)
Comment (AndrewEichmann-NOAA) [reply parm/soca/berror/soca_ensrecenter.yaml]: changed

File: parm/templates/recen.yaml (hunk 6 above)
Comment (AndrewEichmann-NOAA) [reply parm/templates/recen.yaml]: this yaml has been banished

File: ush/soca/marine_recenter.py (hunk 8 above)
Comment (AndrewEichmann-NOAA) [reply ush/soca/marine_recenter.py]: I prefer gdas

File: parm/soca/berror/soca_ensrecenter.yaml (hunk 1 above)
Comment (AndrewEichmann-NOAA) [reply parm/soca/berror/soca_ensrecenter.yaml]: corrected

**Comment by AndrewEichmann-NOAA** [comment]
Haha:
```
      Start 1754: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_ECEN
32/54 Test #1754: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_ECEN ....   Passed  170.83 sec
```

**Comment by AndrewEichmann-NOAA** [comment]
In all seriousness, the ecen produces (almost) what is expected -- I have to remove the `restart` subdir -- so was the CI failing earlier?
```
-bash-4.2$ ls gw/testrun/testjjobs/ROTDIRS/gdas_test/enkfgdas.20180415/12/mem00?/analysis/ocean/restart
gw/testrun/testjjobs/ROTDIRS/gdas_test/enkfgdas.20180415/12/mem001/analysis/ocean/restart:
enkfgdas.t12z.ocninc.nc

gw/testrun/testjjobs/ROTDIRS/gdas_test/enkfgdas.20180415/12/mem002/analysis/ocean/restart:
enkfgdas.t12z.ocninc.nc

gw/testrun/testjjobs/ROTDIRS/gdas_test/enkfgdas.20180415/12/mem003/analysis/ocean/restart:
enkfgdas.t12z.ocninc.nc

gw/testrun/testjjobs/ROTDIRS/gdas_test/enkfgdas.20180415/12/mem004/analysis/ocean/restart:
enkfgdas.t12z.ocninc.nc
-bash-4.2$
```

**Comment by guillaumevernieres** [comment]
I just committed the bugfix to your branch @AndrewEichmann-NOAA (I hope it's the only one). We'll re-run the ci once you fix your code norm issues.

**Comment by guillaumevernieres** [comment]
> In all seriousness, the ecen produces (almost) what is expected -- I have to remove the `restart` subdir -- so was the CI failing earlier?
> 
> ```
> -bash-4.2$ ls gw/testrun/testjjobs/ROTDIRS/gdas_test/enkfgdas.20180415/12/mem00?/analysis/ocean/restart
> gw/testrun/testjjobs/ROTDIRS/gdas_test/enkfgdas.20180415/12/mem001/analysis/ocean/restart:
> enkfgdas.t12z.ocninc.nc
> 
> gw/testrun/testjjobs/ROTDIRS/gdas_test/enkfgdas.20180415/12/mem002/analysis/ocean/restart:
> enkfgdas.t12z.ocninc.nc
> 
> gw/testrun/testjjobs/ROTDIRS/gdas_test/enkfgdas.20180415/12/mem003/analysis/ocean/restart:
> enkfgdas.t12z.ocninc.nc
> 
> gw/testrun/testjjobs/ROTDIRS/gdas_test/enkfgdas.20180415/12/mem004/analysis/ocean/restart:
> enkfgdas.t12z.ocninc.nc
> -bash-4.2$
> ```

Somebody with a first name starting with `G` and ending with `uillaume` did a bad review and also decided to not run the ci on @apchoiCMD 's PR yesterday. Oops. 
Look at the deterministic. The increment doesn't go under `restart`.

**Comment by AndrewEichmann-NOAA** [comment]
> Somebody with a first name starting with `G` and ending with `uillaume` did a bad review and also decided to not run the ci on @apchoiCMD 's PR yesterday. Oops. Look at the deterministic. The increment doesn't go under `restart`.

Yup, caught and changed that

**Comment by AndrewEichmann-NOAA** [comment]
The trick is to avoid doing reviews

File: parm/soca/berror/soca_ensrecenter.yaml [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+  ocn_filename: MOM.res.nc
+  ice_filename: cice.res.nc
+  read_from_file: 1
+
+output increment:
+  datadir: ./
+  date: '{{ ATM_WINDOW_BEGIN }}'
+  exp: trash
+  type: incr
+  output file: 'ocn.pert.steric.%mem%.nc'
```
Comment (guillaumevernieres) [inline parm/soca/berror/soca_ensrecenter.yaml]:
```suggestion
  output file: 'ocn.recenter.incr.%mem%.nc'
```

File: ush/soca/marine_recenter.py [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+            mem_dir = os.path.join(self.config.ROTDIR,
+                                   f'enkf{RUN}.{PDYstr}',
+                                   f'{cyc}',
+                                   f'mem{str(mem).zfill(3)}',
+                                   'analysis',
+                                   'ocean')
+            mem_dir_real = os.path.realpath(mem_dir)
+            mem_dir_list.append(mem_dir_real)
+
+            copy_list.append([f'ocn.pert.steric.{str(mem)}.nc',
```
Comment (guillaumevernieres) [inline ush/soca/marine_recenter.py]:
```suggestion
            copy_list.append([f'ocn.recenter.incr.{str(mem)}.nc',
```

**Comment by AndrewEichmann-NOAA** [comment]
Now when I run the ctests analysis prep failes with this:

```
Traceback (most recent call last):
  File "/scratch1/NCEPDEV/da/Andrew.Eichmann/fv3gfs/novrfy/global-workflow/sorc/gdas.cd/scripts/exgdas_global_marine_analysis_prep.py", line 274, in <module>
    ufsda.yamltools.save_check(varconfig.as_dict(), target=var_yaml, app='var')
  File "/scratch1/NCEPDEV/da/Andrew.Eichmann/fv3gfs/novrfy/global-workflow/sorc/gdas.cd/bundle/gdas/ush/ufsda/yamltools.py", line 37, in save_check
    wxflow.save_as_yaml(config, target)
NameError: name 'wxflow' is not defined
```

**Comment by AndrewEichmann-NOAA** [comment]
Did this get introduced with pulling out the ufsda stuff?

**Comment by guillaumevernieres** [comment]
> Did this get introduced with pulling out the ufsda stuff?

Ha ... Not sure how we forgot about the prep.py dependency on ufsda.

**Comment by AndrewEichmann-NOAA** [comment]
the bugfix is in

**Comment by guillaumevernieres** [comment]
Bummer, it doesn't build anymore!

**Comment by AndrewEichmann-NOAA** [comment]
The log doesnin't indicate what the problem is

**Comment by RussTreadon-NOAA** [comment]
The rebuild worked.
```
tail -f /scratch1/NCEPDEV/da/Cory.R.Martin/CI/GDASApp/workflow/PR/968/global-workflow/sorc/logs/build_gdas.log
[100%] Linking CXX executable ../../../bin/gdas_obsprovider2ioda.x
cd /scratch1/NCEPDEV/da/Cory.R.Martin/CI/GDASApp/workflow/PR/968/global-workflow/sorc/gdas.cd/build/gdas-utils/obsproc/applications && /scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.6.0/envs/unified-env/install/intel/2021.5.0/cmake-3.23.1-tc5whhf/bin/cmake -E remove /scratch1/NCEPDEV/da/Cory.R.Martin/CI/GDASApp/workflow/PR/968/global-workflow/sorc/gdas.cd/build/bin/gdas_obsprovider2ioda.x
cd /scratch1/NCEPDEV/da/Cory.R.Martin/CI/GDASApp/workflow/PR/968/global-workflow/sorc/gdas.cd/build/gdas-utils/obsproc/applications && /scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.6.0/envs/unified-env/install/intel/2021.5.0/cmake-3.23.1-tc5whhf/bin/cmake -E cmake_link_script CMakeFiles/gdas_obsprovider2ioda.x.dir/link.txt --verbose=NO
make[2]: Leaving directory `/scratch1/NCEPDEV/da/Cory.R.Martin/CI/GDASApp/workflow/PR/968/global-workflow/sorc/gdas.cd/build'
[100%] Built target gdas_obsprovider2ioda.x
make[1]: Leaving directory `/scratch1/NCEPDEV/da/Cory.R.Martin/CI/GDASApp/workflow/PR/968/global-workflow/sorc/gdas.cd/build'
/scratch1/NCEPDEV/nems/role.epic/spack-stack/spack-stack-1.6.0/envs/unified-env/install/intel/2021.5.0/cmake-3.23.1-tc5whhf/bin/cmake -E cmake_progress_start /scratch1/NCEPDEV/da/Cory.R.Martin/CI/GDASApp/workflow/PR/968/global-workflow/sorc/gdas.cd/build/CMakeFiles 0
+ cd ../
+ set +x
+ exit
```

**Review by guillaumevernieres — APPROVED** [review]
Looks good.

**Pull request merged**